x-pack/plugins/reporting/server/browsers/chromium/driver/screenshot_stitcher/combine.ts

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// No types found for this package. May want to investigate an alternative with types.
+// @ts-ignore: implicit any for JS file
+import $streamToObservable from '@samverschueren/stream-to-observable';
+import { PNG } from 'pngjs';
+import * as Rx from 'rxjs';
+import { ObservableInput } from 'rxjs';
+import { map, mergeMap, reduce, switchMap, tap, toArray } from 'rxjs/operators';
+import { Logger, Screenshot, Size } from './types';
+
+// if we're only given one screenshot, and it matches the output dimensions
+// we're going to skip the combination and just use it
+const canUseFirstScreenshot = (
+  screenshots: Screenshot[],
+  size: { width: number; height: number }
+) => {
+  if (screenshots.length !== 1) {
+    return false;
+  }
+
+  const firstScreenshot = screenshots[0];
+  return (
+    firstScreenshot.dimensions.width === size.width &&
+    firstScreenshot.dimensions.height === size.height
+  );
+};
+
+/**
+ * Combines the screenshot clips into a single screenshot of size `outputDimensions`.
+ * @param screenshots - Array of screenshots to combine
+ * @param outputSize - Final output size that the screenshots should match up with
+ * @param logger - logger for extra debug output
+ */
+export function $combine(
+  screenshots: Screenshot[],
+  outputSize: Size,
+  logger: Logger
+): Rx.Observable<string> {
+  logger.debug(
+    `Combining screenshot clips into final, scaled output dimension of ${JSON.stringify(
+      outputSize
+    )}`
+  );
+
+  if (screenshots.length === 0) {
+    return Rx.throwError('Unable to combine 0 screenshots');
+  }
+
+  if (canUseFirstScreenshot(screenshots, outputSize)) {
+    return Rx.of(screenshots[0].data);
+  }
+
+  // Turn the screenshot data into actual PNGs
+  const pngs$ = Rx.from(screenshots).pipe(
+    mergeMap(
+      (screenshot: Screenshot): ObservableInput<PNG> => {
+        const png = new PNG();
+        const buffer = Buffer.from(screenshot.data, 'base64');
+        const parseAsObservable = Rx.bindNodeCallback(png.parse.bind(png));
+        return parseAsObservable(buffer);
+      },
+      (screenshot: Screenshot, png: PNG) => {
+        if (
```

Comment:
This is new too, this is to detect earlier the bug in #19563.  I simulated an error to verify it is exposed correctly:

<img width="1190" alt="screen shot 2018-06-20 at 9 41 50 am" src="https://user-images.githubusercontent.com/16563603/41662541-64cd875a-746f-11e8-9929-52c00b4c00f5.png">
 
(Note after that screenshot I made slight adjustments to the message).